File: src/Client.UWP/Domain/WaterTextures.cs
using Client.View;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace Client.Domain
{
    /// <summary>
    /// Extracts and holds textures for Water map
    /// </summary>
    /// <remarks>
    /// All textures are hold in one big texture, so need to have separated logic which knowledge
    /// where are located particular types of some terrain type.
    /// In that case, <see cref="WaterTextures"/> is responsible to load water tiles located in
    /// some known places in big texture.
    /// </remarks>
    public sealed class WaterTextures
    {
        public TextureHolder CoastWithLandToTheNorth { get; private set; } = new TextureHolder();
        public TextureHolder CoastWithLandToTheSouth { get; private set; } = new TextureHolder();
        public TextureHolder CoastWithLandToTheWest { get; private set; } = new TextureHolder();
        public TextureHolder CoastWithLandToTheEast { get; private set; } = new TextureHolder();
        public TextureHolder CoastWithLandToTheNorthEast { get; private set; } = new TextureHolder();
        public TextureHolder CoastWithLandToTheNorthWest { get; private set; } = new TextureHolder();
        public TextureHolder CoastWithLandToTheSouthEast { get; private set; } = new TextureHolder();
        public TextureHolder CoastWithLandToTheSouthWest { get; private set; } = new TextureHolder();
        public TextureHolder Sea { get; private set; } = new TextureHolder();

        public WaterTextures()
        {
        }

        public WaterTextures(Texture2D terrain)
        {
            var spriteSize = new Point(32, 32);
            CoastWithLandToTheNorth = new TextureHolder(terrain, new Rectangle(new Point(10 * 32, 0 * 32), spriteSize));
            CoastWithLandToTheSouth = new TextureHolder(terrain, new Rectangle(new Point(4 * 32, 0 * 32), spriteSize));
            CoastWithLandToTheWest = new TextureHolder(terrain, new Rectangle(new Point(6 * 32, 0 * 32), spriteSize));
            CoastWithLandToTheEast = new TextureHolder(terrain, new Rectangle(new Point(18 * 32, 0 * 32), spriteSize));
            CoastWithLandToTheNorthEast = new TextureHolder(terrain, new Rectangle(new Point(8 * 32, 0 * 32), spriteSize));
            CoastWithLandToTheNorthWest = new TextureHolder(terrain, new Rectangle(new Point(0 * 32, 1 * 32), spriteSize));
            CoastWithLandToTheSouthEast = new TextureHolder(terrain, new Rectangle(new Point(14 * 32, 0 * 32), spriteSize));
            CoastWithLandToTheSouthWest = new TextureHolder(terrain, new Rectangle(new Point(6 * 32, 1 * 32), spriteSize));
            Sea = new TextureHolder(terrain, new Rectangle(new Point(2 * 32, 0 * 32), spriteSize));
        }
    }
}

File: core/src/com/mygdx/game/resources/WaterTextures.java
package com.mygdx.game.resources;

import com.badlogic.gdx.math.Rectangle;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.mygdx.game.Config;
import com.mygdx.game.runtime.GameComponentBase;
import com.mygdx.game.view.IntRectangle;
import com.mygdx.game.view.TextureHolder;

import lombok.Getter;

/**
 *  Extracts and holds textures for Water map
 *
 *  All textures are hold in one big texture, so need to have separated logic which knowledge
 *  where are located particular types of some terrain type.
 *  In that case, {@see WaterTextures> is responsible to load water tiles located in
 *  some known places in big texture.
 * 
 */
@Singleton
@Getter
public class WaterTextures extends GameComponentBase {

    private TerrainTexture terrainTexture;
    
    @Inject
    public WaterTextures(TerrainTexture terrainTexture) {
        this.terrainTexture = terrainTexture;
    }

    private TextureHolder landNorth;
    private TextureHolder landSouth;
    private TextureHolder landWest;
    private TextureHolder landEast;
    private TextureHolder landSouthEast;
    private TextureHolder landSouthWest;
    private TextureHolder landNorthEast;
    private TextureHolder landNorthWest;
    private TextureHolder sea;
    private TextureHolder blank;

    @Override
    public void useTextures() {
        var texture = terrainTexture.getTexture();
        var ss = Config.SpriteSize;
        landNorth = new TextureHolder(texture, IntRectangle.of(4 * ss, 0 * ss, ss, ss));
        landSouth = new TextureHolder(texture, IntRectangle.of(10 * ss, 0 * ss, ss, ss));
        landEast = new TextureHolder(texture, IntRectangle.of(6 * ss, 0 * ss, ss, ss));
        landWest = new TextureHolder(texture, IntRectangle.of(18 * ss, 0 * ss, ss, ss));
        landSouthEast = new TextureHolder(texture, IntRectangle.of(14 * ss, 0 * ss, ss, ss)); 
        landSouthWest = new TextureHolder(texture, IntRectangle.of(6 * ss, 1 * ss, ss, ss));
        landNorthEast = new TextureHolder(texture, IntRectangle.of(8 * ss, 0 * ss, ss, ss));
        landNorthWest = new TextureHolder(texture, IntRectangle.of(0 * ss, 1 * ss, ss, ss)); 
        sea = new TextureHolder(texture, IntRectangle.of(2 * ss, 0 * ss, ss, ss));
        blank = new TextureHolder(texture, IntRectangle.of(15 * ss, 8 * ss, ss, ss));
    }
}
